### ISSUE TITLE
[rllib] Keeping the LSTM state across episodes?

### ISSUE THREAD

**Issue by alversafa** [ray-project/ray#6610]
Is there a smooth way to keep the LSTM state across episodes while training PPO with LSTM? Or do we have to alter the source code of the PPO training loop for this? If so, is it possible for you to point me to the place to make this?

**Comment by ericl** [comment]
Yep, you can use soft_horizon for this:
```
    # Calculate rewards but don't reset the environment when the horizon is
    # hit. This allows value estimation and RNN state to span across logical
    # episodes denoted by horizon. This only has an effect if horizon != inf.
    "soft_horizon": False,
```

**Comment by ericl** [comment]
The `soft_horizon` options is close to this, but it currently only applies if the `horizon` option is used to split up episodes.

However I think if you change this line here: https://github.com/ray-project/ray/blob/master/rllib/evaluation/sampler.py#L475

From `if hit_horizon and soft_horizon:` to just `if soft_horizon:`, that would work. Try that out?

**Comment by alversafa** [comment]
Just to make it clear I'm trying to implement the RL^2 paper (https://arxiv.org/pdf/1611.02779.pdf) and according to this paper the state of the LSTM has to be kept across episodes in a trial (which consists of two episodes).

I will try what you suggested. However, it says that setting `soft_horizon=True` will not reset the environment. I think that this may be a problem as the environment has to reset at the end of each episode according to the paper. 

So is there a way to keep the state and also reset the environment at the end of episodes?

**Comment by ericl** [comment]
I see, then in addition I think you also need to change this other line to:
```
                resetted_obs = base_env.try_reset(env_id)
```

so that the env is reset on done, but the episode object is only soft reset.

**Comment by alversafa** [comment]
Thanks, that seems to solve the problem. 

However, I guess I have edited the above comment just before you posted this solution. So according to the above comment I want to keep the LSTM state for says 2 (or 3) episodes and then reset it to zero. Can this also easily be done?

**Comment by ericl** [comment]
>  So according to the above comment I want to keep the LSTM state for says 2 (or 3) episodes and then reset it to zero. Can this also easily be done?

That seems harder to do... I would probably go with modifying the env to internally contain multiple episodes instead, or something like that. You'd have to somehow account for the rewards separately then, since RLlib couldn't do it for you, but it could be perhaps exposed as a custom metric.

**Comment by alversafa** [comment]
I thought the same way and did modify the environment to have multiple sequences. However, as you told the rewards became a problem. Is it possible for you to give more details on how this can be overcome with custom metrics?

**Comment by ericl** [comment]
I would check out https://ray.readthedocs.io/en/latest/rllib-training.html#callbacks-and-custom-metrics

**Comment by alversafa** [comment]
Another alternative way would be to prevent the call of `model.get_initial_state()` for some number of `done=True` in the main training loop of PPO. However, I am having a hard time in finding out where exactly this happens. Do you have any idea on this?

**Comment by ericl** [comment]
Should be this line here: https://github.com/ray-project/ray/blob/master/rllib/evaluation/episode.py#L149

Which is called in a couple places in the sampler: https://github.com/ray-project/ray/blob/master/rllib/evaluation/sampler.py#L420
